### PR TITLE
Add Thanos Ruler

### LINKED ISSUE #784: [3] Remove wc-reader-prometheus
**Is your feature request related to a problem? Please describe.**

Since we would like to remove wc-reader-prometheus we somehow need to get wc alerts somehow. One solution would be to use the rule component and send alerts to our current alertmanager. Another solution would be to setup an ops alertmanager in wc.

Investigate a little bit and see what's the best option. Take at least this into consideration https://thanos.io/v0.24/components/rule.md/#risk

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

**Definition of done:**
When could this feature be considered done.

* WC-reader-prometheus is gone
* wc alerts are going to ops alertmanager

### PR DESCRIPTION
**Pull request by aarnq**
**What this PR does / why we need it**:
This enables us to use thanos ruler to evaluate alert rules through thanos, allowing us to remove the wc-reader and the last blocker to finally remove influxdb.

**Which issue this PR fixes**:
fixes #784

**Public facing documentation PR** *(if applicable)*

**Special notes for reviewer**:
I re-enabled one metric that is quite big, I will monitor it a bit to see. It's related to the kube-apiserver showing its error budget (dashboard + alert). I don't know how important that metric is to us or if I just should drop it.

(I will retest to see how it is to apply over main)

**Checklist:**

- [x] Added relevant notes to [WIP-CHANGELOG.md](https://github.com/elastisys/compliantkubernetes-apps/blob/main/WIP-CHANGELOG.md)
- [x] Proper commit message prefix on all commits
- [ ] Updated the [public facing documentation](https://github.com/elastisys/compliantkubernetes)
- Is this changeset backwards compatible for existing clusters? Applying:
    - [ ] is completely transparent, will not impact the workload in any way.
    - [ ] requires running a migration script.
    - [ ] will create noticeable cluster degradation.
          E.g. logs or metrics are not being collected or Kubernetes API server
          will not be responding while upgrading.
    - [ ] requires draining and/or replacing nodes.
    - [ ] will change any APIs.
          E.g. removes or changes any CK8S config options or Kubernetes APIs.
    - [ ] will break the cluster.
          I.e. full cluster migration is required.
- Chart checklist (pick exactly one):
    - [x] I upgraded no Chart.
    - [ ] I upgraded a Chart and determined that no migration steps are needed.
    - [ ] I upgraded a Chart and added [migration steps](https://github.com/elastisys/compliantkubernetes-apps/blob/main/migration).

**Pipeline config** *(if applicable)*
If you change some config options (e.g. add/rename variable or change the default value) you may need to update the config used by the pipeline in `pipeline/config`.

### CODE REVIEW

File: helmfile/charts/grafana-ops/dashboards/gatekeeper-dashboard.json [hunk 1]
```diff
@@ -20,7 +20,7 @@
   "links": [],
   "panels": [
     {
-      "content": "\n# Detailed gatekeeper logs can be found in kibana:\n\n\n<<kibanaURL>>",
+      "content": "\n# Detailed gatekeeper logs can be found in OpenSearch Dashboards:\n\n\n<<kibanaURL>>",
```
Comment (robinAwallace) [inline helmfile/charts/grafana-ops/dashboards/gatekeeper-dashboard.json:23]: \n\n\n<<kibanaURL>>
Does kibanaURL matter here?

**Review by robinAwallace — COMMENTED** [review]
LGTM :slightly_smiling_face:

File: helmfile/charts/grafana-ops/dashboards/gatekeeper-dashboard.json (hunk 1 above)
Comment (aarnq) [reply helmfile/charts/grafana-ops/dashboards/gatekeeper-dashboard.json]: Fixed it :slightly_smiling_face:

**Comment by crssnd** [comment]
How will we manage the alerts in prometheus-alerts when we will upgrade kube-prometheus-stack upstream chart?

**Comment by aarnq** [comment]
> How will we manage the alerts in prometheus-alerts when we will upgrade kube-prometheus-stack upstream chart?

I think we have to investigate that a bit. We could use the upstream of kube-prometheus-stack and generate rules from the kubernetes-mixins project with the changes we require, and then split on record and alert rules.

File: helmfile/values/kube-prometheus-stack-sc.yaml.gotmpl [hunk 2]
```diff
@@ -25,7 +16,7 @@ kubeApiServer:
     metricRelabelings:
       - action: keep
         sourceLabels: [__name__]
-        regex: '(process_cpu_seconds_total|process_resident_memory_bytes|go_goroutines|workqueue_queue_duration_seconds_bucket|workqueue_adds_total|workqueue_depth|apiserver_request_total|apiserver_request_count|apiserver_request_latencies_bucket|apiserver_client_certificate_expiration_seconds_count|apiserver_request_duration_seconds_sum|apiserver_request_duration_seconds_count|APIServiceRegistrationController_depth)'
+        regex: '(process_cpu_seconds_total|process_resident_memory_bytes|go_goroutines|workqueue_queue_duration_seconds_bucket|workqueue_adds_total|workqueue_depth|apiserver_request_total|apiserver_request_count|apiserver_request_duration_seconds_bucket|apiserver_request_latencies_bucket|apiserver_client_certificate_expiration_seconds_count|apiserver_request_duration_seconds_sum|apiserver_request_duration_seconds_count|APIServiceRegistrationController_depth)'
```
[the remark this answers is not in the document]
Comment (aarnq) [reply helmfile/values/kube-prometheus-stack-sc.yaml.gotmpl:19]: Then I'll go ahead and remove it.

**Comment by crssnd** [comment]
> > How will we manage the alerts in prometheus-alerts when we will upgrade kube-prometheus-stack upstream chart?
> 
> I think we have to investigate that a bit. We could use the upstream of kube-prometheus-stack and generate rules from the kubernetes-mixins project with the changes we require, and then split on record and alert rules.

ok please create another issue for that, if you cannot do it as part of this PR
also add some notes in the [upstream Readme](https://github.com/elastisys/compliantkubernetes-apps/blob/main/helmfile/upstream/README.md) so we know to update the rules when we update the kube-prometheus-stack chart

File: helmfile/charts/prometheus-alerts/templates/alerts/alertmanager.rules.yaml [hunk 3]
```diff
@@ -27,10 +28,10 @@ spec:
     - alert: AlertmanagerConfigInconsistent
       annotations:
         message: The configuration of the instances of the Alertmanager cluster `{{`{{`}}$labels.service{{`}}`}}` are out of sync.
-      expr: count_values("config_hash", alertmanager_config_hash{job="{{ $alertmanagerJob }}",namespace="{{ $namespace }}"}) BY (service) / ON(service) GROUP_LEFT() label_replace(max(prometheus_operator_spec_replicas{job="{{ $operatorJob }}",namespace="{{ $namespace }}",controller="alertmanager"}) by (name, job, namespace, controller), "service", "$1", "name", "(.*)") != 1
+      expr: count_values("config_hash", alertmanager_config_hash{job="{{ $alertmanagerJob }}",namespace="{{ $namespace }}"}) BY (cluster, service) / ON(cluster, service) GROUP_LEFT() label_replace(max(prometheus_operator_spec_replicas{job="{{ $operatorJob }}",namespace="{{ $namespace }}",controller="alertmanager"}) by (cluster, name, job, namespace, controller), "service", "$1", "name", "(.*)") != 1
       for: 5m
       labels:
-        severity: critical
+        severity: criticalExemplars
```
Comment (viktor-f) [inline helmfile/charts/prometheus-alerts/templates/alerts/alertmanager.rules.yaml:34]: Why is this changed?

**Comment by aarnq** [comment]
Updated and checked against #849. Will clean and squash commits once it is in.

File: helmfile/50-applications.yaml [hunk 4]
```diff
@@ -515,7 +515,7 @@ releases:
     app: thanos-ruler
   chart: ./charts/thanos/ruler
   version: 0.1.0
-  installed: {{ and .Values.thanos.enabled .Values.thanos.ruler.enabled }}
+  installed: {{ .Values.thanos.ruler.enabled }}
```
Comment (robinAwallace) [inline helmfile/50-applications.yaml:518]: Should the ruler not be disabled if you disable thanos?

Comment (aarnq) [reply helmfile/50-applications.yaml:518]: With #849 the option to disable Thanos will be gone.

**Review by crssnd — APPROVED** [review]

File: WIP-CHANGELOG.md [hunk 5]
```diff
@@ -49,8 +49,17 @@
...
+   - Modified rules to allow for different labels for alert and record rules, and to pass the cluster label through aggregations
+   - Unused rules have been dropped
+ - Grouped thanos charts
+ - Configured thanos-ruler so it is enabled by default, runs as an HA pair without persistence, and dynamically reloads its rules on changes
+ - Changed dashboards previously defaulting to wc-reader to default to "Thanos All"
+ - Changed service cluster prometheus to use an external label instead of service monitor relabeling and write relabeling
 
 ### Fixed
+- Fixed ceph cluster dashboard to retrieve the cluster label from ceph metrics
+- Fixed opensearch naming on gatekeeper dashboard
```
Comment (OlleLarsson) [inline WIP-CHANGELOG.md:63]:
```suggestion
```

File: config/config/sc-config.yaml [hunk 6]
```diff
@@ -349,6 +351,11 @@ thanos:
         cpu: 300m
         memory: 100Mi
 
+    # With our configuration of rules thanos ruler has no recording rules so it should be fine without persistence.
```
Comment (OlleLarsson) [inline config/config/sc-config.yaml:354]:
```suggestion
    # With our configuration of rules, thanos ruler has no recording rules so it should be fine without persistence.
```

**Review by Xartos — APPROVED** [review]
LGTM

**Review by robinAwallace — APPROVED** [review]

**Review by OlleLarsson — APPROVED** [review]
:tada:
